src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +499,137 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(void *address, unsigned *is_vmm,
+                         ucs_memory_type_t *vmm_mem_type)
+{
+    ucs_status_t status        = UCS_OK;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop   = {};
+    CUmemGenericAllocationHandle alloc_handle;
+#endif
+
+    *is_vmm = 0;
+
+#if HAVE_CUDA_FABRIC
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemRetainAllocationHandle(&alloc_handle, (void*)address));
+    if (status != UCS_OK) {
+        status = UCS_OK;
+        goto out;
+    }
+
+    *is_vmm = 1;
+    status  = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle));
+    if (status != UCS_OK) {
+        status = UCS_ERR_INVALID_ADDR;
```

Comment:
1. why set status to a different error code?
2. is it ok to examine the value of prop in lines 531-539 even if we have failure in line 526? seems wrong

Reply:
Thanks for pointing out the mistake. There is a return status required after line 528 here. I've addressed this now.